derive_miniconf/src/lib.rs

```diff
@@ -49,162 +37,157 @@ impl TypeDefinition {
-                    state[0] += 1;
-
-                    return Ok(true);
-                }
+    } else {
+        quote! {
+            #i => {
+                path.push_str(stringify!(#field_name))
+                    .map_err(|_| miniconf::IterError::PathLength)?;
+                state[0] += 1;
+
+                return Ok(true);
             }
         }
-    });
+    }
+}
 
-    let metadata_arms = fields.iter().enumerate().map(|(i, f)| {
-        let field_type = &f.field.ty;
-        let field_name = &f.field.ident;
-        if f.deferred {
-            quote! {
-                #i => {
-                    let mut meta = <#field_type>::metadata();
+fn metadata_arm((i, f): (usize, &StructField)) -> proc_macro2::TokenStream {
+    let field_type = &f.field.ty;
+    let field_name = &f.field.ident;
+    if f.deferred {
+        quote! {
+            #i => {
+                let mut meta = <#field_type>::metadata();
 
-                    // Unconditionally account for separator since we add it
-                    // even if elements that are deferred to (`Options`)
-                    // may have no further hierarchy to add and remove the separator again.
-                    meta.max_length += concat!(stringify!(#field_name), "/").len();
-                    meta.max_depth += 1;
+                // Unconditionally account for separator since we add it
+                // even if elements that are deferred to (`Options`)
+                // may have no further hierarchy to add and remove the separator again.
+                meta.max_length += stringify!(#field_name).len() + 1;
+                meta.max_depth += 1;
 
-                    meta
-                }
+                meta
             }
-        } else {
-            quote! {
-                #i => {
-                    let mut meta = miniconf::Metadata::default();
+        }
+    } else {
+        quote! {
+            #i => {
+                let mut meta = miniconf::Metadata::default();
 
-                    meta.max_length = stringify!(#field_name).len();
-                    meta.max_depth = 1;
-                    meta.count = 1;
+                meta.max_length = stringify!(#field_name).len();
+                meta.max_depth = 1;
+                meta.count = 1;
 
-                    meta
-                }
+                meta
             }
         }
-    });
+    }
+}
 
-    let (impl_generics, ty_generics, where_clause) = typedef.generics.split_for_impl();
-    let name = typedef.name;
+/// Derive the Miniconf trait for structs.
+///
+/// # Args
+/// * `input` - The derive input token stream.
+/// * `data` - The data associated with the struct definition.
+/// * `atomic` - specified true if the data must be updated atomically. If false, data must be
+///   set at a terminal node.
+///
+/// # Returns
+/// A token stream of the generated code.
+fn derive_struct(mut input: DeriveInput) -> TokenStream {
```

Comment:
Why take in the `input` and then desugar it into `syn::Data::Struct(data)` again after we already do this at the top level?

Reply:
In `derive()` we're only checking for `data` variant to dispatch to the right function.
Here we also need other fields (other than `data`).

Reply:
I don't see a clear alternative.

Reply:
You could just destructure in `derive` and immediately pass in `generics` and `name` as input args to `derive_struct`

Reply:
I fear that would be more code than the match

Reply:
I don't think so? I think it's just two additional args for the function for `input.ident` and `input.generics`?